server/src/main.rs

```diff
@@ -106,76 +108,90 @@ fn main() {
+                .expect("failed to instrument logging and tracing!");
+
+            // Log panics as tracing errors if possible
+            let prev_hook = std::panic::take_hook();
+            std::panic::set_hook(Box::new(move |panic_info| {
+                tracing_panic::panic_hook(panic_info);
+                // run original hook if any.
+                prev_hook(panic_info);
+            }));
+
+            info!("Starting Restate Server {}", build_info::build_info());
+            info!(
+                "Loading configuration file from {}",
+                cli_args.config_file.display()
+            );
+            info!(
+                "Configuration dump (MAY CONTAIN SENSITIVE DATA!):\n{}",
+                serde_yaml::to_string(&config).unwrap()
+            );
+
+            WipeMode::wipe(
+                cli_args.wipe.as_ref(),
+                config.node.meta.storage_path().into(),
+                config.node.worker.storage_path().into(),
+            )
+            .await
+            .expect("Error when trying to wipe the configured storage path");
+
+            let task_center_watch = tc.watch_shutdown();
+            let node = Node::new(config.node);
+            if let Err(err) = node {
+                handle_error(err);
             }
+            // We ignore errors since we will wait for shutdown below anyway.
+            // This starts node roles and the rest of the system async under tasks managed by
+            // the TaskCenter.
+            let _ = node.unwrap().boot(&tc);
+
+            tokio::select! {
+                signal_name = signal::shutdown() => {
+                    info!("Received shutdown signal.");
+                    let signal_reason = format!("received signal {}", signal_name);
+
+                    let shutdown_with_timeout = tokio::time::timeout(
+                        config.shutdown_grace_period.into(),
+                        tc.shutdown_node(&signal_reason, 0)
+                    );
+
+                    // ignore the result because we are shutting down
+                    let shutdown_result = shutdown_with_timeout.await;
+
+                    if shutdown_result.is_err() {
+                        warn!("Could not gracefully shut down Restate, terminating now.");
+                    } else {
+                        info!("Restate has been gracefully shut down.");
+                    }
+                },
+                _ = task_center_watch => {
+                    // Shutdown was requested by task center and it has completed.
+                },
+            };
+
+            shutdown_tracing(config.shutdown_grace_period.div(2), tracing_guard).await;
         }
-
-        shutdown_tracing(config.shutdown_grace_period.div(2), tracing_guard).await;
     });
+    let exit_code = tc.exit_code();
+    if exit_code != 0 {
+        error!("Restate terminated with exit code {}!", exit_code);
+    } else {
+        info!("Restate terminated");
+    }
+    drop(runtime);
```

Comment:
This is not necessary but I left it here to as a visual indicator that we will drop tokio before exiting. This can be removed in the future.